File: 프로그래머스/2/12952. N－Queen/N－Queen.java
public class Solution {

    public static int[] queens;
    public static int answer = 0;

    public static int solution(int n) {
        queens = new int[n];
        dfs(0);
        return answer;
    }

    public static void dfs(int depth) {
        if (depth == queens.length) {
            answer++;
            return;
        }

        for (int i = 0; i < queens.length; ++i) {
            queens[depth] = i;

            if (isAvailable(depth)) {
                dfs(depth + 1);
            }
        }
    }

    public static boolean isAvailable(int depth) {
        for (int i = 0; i < depth; ++i) {
            if (queens[i] == queens[depth] || Math.abs(queens[i] - queens[depth]) == Math.abs(i - depth)) {
                return false;
            }
        }
        return true;
    }

}

File: 프로그래머스/2/12952. N－Queen/N－Queen.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

bool check(vector<int> m, int depth)
{
    for (int i = 0; i < depth; ++i)
    {
        if (m[i] == m[depth])
            return false;

        if (abs(m[i] - m[depth]) == abs(i - depth))
            return false;
    }

    return true;
}

void dfs(vector<int> m, int depth, int n)
{
    if (depth >= n)
    {
        ++answer;
        return;
    }

    for (int i = 0; i < n; ++i)
    {
        m[depth] = i;

        if (check(m, depth) == true)
        {
            dfs(m, depth + 1, n);
        }
    }
}


int solution(int n) {

    dfs(vector<int>(n), 0, n);

    return answer;
}